Feature Report: GLVTX

jroweboy opened the pull request:
This is the rough draft of the new blog post. I'm putting this up now so people can read it and find any oddities early. I'm mostly looking for feedback on style, but I'll deal with minor grammar fixes. Remember this is a rough draft so now would be the time to nitpick it to death before I consider the core content finalized

Please fact check the article as well, especially when I get the "Challenges of PICA -> GPU" part written.

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+
+## Cut to the chase, how fast is this?
+
+Very. Test results across various computers show that it averages out to be a **2x speed boost for almost everyone.**
```
Using average and almost in the same sentence makes the statement weak. I would take out "for almost everyone" considering there's a wide variance.

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+gif: side by side pokemon before and after
+subtext: Real time footage with the frame limiter off. Left side is before this change and right side is after.
+
+With the new update, Citra will use much more of your GPU, removing some of the dependence on having a CPU with high single-core performance. As always, the actual difference will vary by game and by your hardware.
```
removing some of -> lessening

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+
+## History of Citra's Rendering Backends
+
+Back in early 2015, Citra was still a young emulator, and just barely started displaying graphics for the first time. In a momentous occasion, Citra displayed 3D graphics from a commercial game, Legend of Zelda: Ocarina of Time 3D
```
and had just

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+video: https://www.youtube.com/watch?v=Hj8sPsB5qXQ 
+subtext: Good thing that shield flashing bug still isn't around
+
+How is there such a major difference in speed from the first and the second video? The speed difference boils down to how the 3DS GPU, the PICA200, is being emulated. The first video is showing off the software renderer, which emulates the PICA200 by using your host computer's CPU; on the other hand, the second video is using the OpenGL hardware renderer, which emulates the PICA200 by using your  ...
```
Maybe combine the two sentences about using the CPU and GPU somehow

FearlessTobi replied:
it's

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+video: https://www.youtube.com/watch?v=sYukdJam6gk
+subtext: It took a year from when citra started to get to this point
+
+This engineering feat was thanks to the hardwork of many contibutors in both the emulator scene and the 3ds hacking scene, who worked tirelessly to reverse engineer the 3DS GPU, a chip called the PICA200. But not even a few months later, Citra was able to play the game at fullspeed! 
```
Maybe mention the chip's obscurity?

FearlessTobi commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+picture: pokemon outline bug
+subtext: The infamous pokemon outlines bug was one such example of something that OpenGL just doesn't support
+
+When the hardware renderer was originally written, there was no other choice to use besides OpenGL. On the surface, OpenGL is a good candidate for a hardware renderer, as it runs on all 3 major desktop platforms. As long as we stay with the [8 year old OpenGL 3.3 standard](https://www.khronos.org/opengl/wiki/History_of_OpenGL#OpenGL_3.3_.282010.29), we can support a very wide range of hardware as ...
```
ditto

BreadFish64 replied:
> Its performance and 

not here

FearlessTobi replied:
"Its not" should be corrected anyway

FearlessTobi commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+picture: metroid samus returns before and after
+subtext: Metroid samus returns heavily used a PICA200 feature that had a `// TODO: implement this` comment in the hardware renderer. Implementing that fixed upscaling while making it much faster
+
+Whenever a texture is used in the hardware renderer, the hardware renderer will try to use a copy of the texture already in the GPU memory, but if that fails, it has to reload the texture from the emulated 3DS memory. This is called a texture upload, and its slow for a good reason. The communication between CPU and GPU is optimized for large amounts of data transferred, but as a tradeoff, its not ...
```
ditto

FearlessTobi commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+picture: pokemon outlines working
+subtext: The long standing issue is gone!
+
+The texture forwarding rewrite has been battle tested in Citra Canary for the last 2 months, during which time we fixed over 20 reported issues. We happy to announce that its now merged into the master branch, so please enjoy the new feature in the latest nightly build!
```
ditto

j-selby replied:
`we`

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+picture: pokemon outline bug
+subtext: The infamous pokemon outlines bug was one such example of something that OpenGL just doesn't support
+
+When the hardware renderer was originally written, there was no other choice to use besides OpenGL. On the surface, OpenGL is a good candidate for a hardware renderer, as it runs on all 3 major desktop platforms. As long as we stay with the [8 year old OpenGL 3.3 standard](https://www.khronos.org/opengl/wiki/History_of_OpenGL#OpenGL_3.3_.282010.29), we can support a very wide range of hardware as ...
```
This whole paragraph makes it sound like we're getting rid of OpenGL, might make people confused and think we're switching to Vulkan. I also don't really see how the MacOS memory leak is related

jroweboy replied:
The ending sentence is why "For a little bit of good news, this is fixed in the latest nightly, but only because the entire texture handling code was rewritten!"

Which leads into the next paragraph about the texture cache rewrite

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+
+## Moving Forward with the Hardware Renderer: Cleaning up Texture Forwarding
+
+Despite the issues mentioned above, OpenGL has been a good choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making it better. Their first major contribution was a massive, [complete rewrite](https://github.com/citra-emu/citra/pull/3281) of the [texture forwarding support](https://citra-emu.org/entry/texture-forwarding-brings-hd-output-to-citra/) th ...
```
there's always a better word than 'good'

BreadFish64 replied:
>OpenGL has been a good choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making it better.

makes it sound like Phantom is improving OpenGL itself

BreadFish64 replied:
forwa**r**ding

BreadFish64 replied:
>The new texture forwaring code increases the speed of many games, while also fixing upscaled rendering in some other games.

shifts from present to present progressive

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+
+Despite the issues mentioned above, OpenGL has been a good choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making it better. Their first major contribution was a massive, [complete rewrite](https://github.com/citra-emu/citra/pull/3281) of the [texture forwarding support](https://citra-emu.org/entry/texture-forwarding-brings-hd-output-to-citra/) th ...
+
+picture: metroid samus returns before and after
```
capitalize game names

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+Despite the issues mentioned above, OpenGL has been a good choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making it better. Their first major contribution was a massive, [complete rewrite](https://github.com/citra-emu/citra/pull/3281) of the [texture forwarding support](https://citra-emu.org/entry/texture-forwarding-brings-hd-output-to-citra/) th ...
+
+picture: metroid samus returns before and after
+subtext: Metroid samus returns heavily used a PICA200 feature that had a `// TODO: implement this` comment in the hardware renderer. Implementing that fixed upscaling while making it much faster
```
replace 'that' with 'it' maybe, and 'it' with 'the game'

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+
+Whenever a texture is used in the hardware renderer, the hardware renderer will try to use a copy of the texture already in the GPU memory, but if that fails, it has to reload the texture from the emulated 3DS memory. This is called a texture upload, and its slow for a good reason. The communication between CPU and GPU is optimized for large amounts of data transferred, but as a tradeoff, its not ...
+
+The texture forwarding rewrite increases the speed of many games by adding in new checks to avoid this costly syncronization of textures between emulated 3DS memory and the host GPU memory. Additionally, the new texture forwarding can avoid even more texture uploads by copying the data from an existing and compatible location. As an extension of this feature, phantom went the extra mile and fixed ...
```
had to what?

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,95 @@
+
+## There's Always More To Do
+
+While today marks a victory for fast emulation, we always have room for improvement. As explained earlier in the article, getting OpenGL to work consistently across all platforms and GPUs is surprisingly challenging, so be ready for bugs. Additionally, there is always something more that can be done to make the hardware renderer even faster and more accurate. But in the meantime, we hope you enjo ...
```
>Additionally, there is always something more that can be done to make the hardware renderer even faster and more accurate.

'additionally' makes it sound like you're gonna give a specific example

j-selby commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,133 @@
+
+## Moving Forward with the Hardware Renderer: Cleaning up Texture Forwarding
+
+Despite the issues mentioned above, OpenGL has been a good choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making it better. Their first major contribution was a massive, [complete rewrite](https://github.com/citra-emu/citra/pull/3281) of the [texture forwarding support](https://citra-emu.org/entry/texture-forwarding-brings-hd-output-to-citra/) th ...
```
`Forwaring` => `Forwarding`

j-selby commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,133 @@
+
+#### Geometry shaders are back with a vengence
+
+The veteren Citra user will remember a time long ago when Citra didn't support Geometry Shaders. It was a dark time where [Fire Emblem was mostly unplayable](https://github.com/citra-emu/citra/issues/2548) and [Pokemon didn't have attack animations](https://github.com/citra-emu/citra/issues/2629). Geometry shaders is a GPU stage where the game can transform a single primitive into zero or more pr ...
```
`veteren`

JMC47 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -85,11 +87,47 @@ graph: performance of low end gpus
+
+We were very excited to launch this feature when phantom declared that it was ready; results from user testing was entirely positive, and the performance improvements were unbelievable, but one thing stood in the way. No one had yet tested to see if it worked on AMD GPUs. We called for our good friend [JMC4789]() to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! ...
+
+From JMC4789's time in Dolphin, he's made a few friends here and there, and so he reached out to a contact that works on AMD drivers. After a few grueling weeks, the driver developer was able to narrow down what the problem is, and surprise, its not a bug in citra, nor it is a bug in the AMD drivers. Its a bug in the GL specification, or more precisely, the issue is ambiguous wording. [glDrawRang ...
```
I'd rather not say I directly contacted the team.  I think a better way to say it would be that I was able to use connections in Dolphin.  I could ask the person I was contacting if they'd mind being mentioned directly as well.

Also, from what I've learned, what Citra was doing was iffy at best, so, calling out AMD on this is not really the right way to go.  I think this is a case where NVIDIA errs on the side of permissive, and AMD errs on teh side of restrictive.

jroweboy commented:
Okay, the rough draft should be complete. I know i left an incomplete sentence in there about geo shader support. This blog post is now compilable, so please preview it there as well.

@bunnei and @phanto-m please take some time to check for technical accuracy. I have a stub in place for some technical comments that I could use some help writing. Also bunnei, imma need a banner unless you would rather have me make one.

@JMC47 I rewrote the AMD section and removed references to the driver team. I also tried to make it sound more lenient to AMD, if you don't like my wording, please suggest something better because I am a programmer not a writer :)

@chris062689 @j-selby i added two new shortcodes. One lets you put images or video side by side, the other is for mp4 video. I also made the figure shortcode accept youtube videos. Pls review.

wwylele commented:
regarding the pipeline chart, the "GLSL" and "GPU" are too vague. I didn't read phantom's code carefully, but at least based on a reasonable setup:
PICA Vertex loader -> GL vertex loader + vertex shader
PICA Vertex shader -> CPU PICA->GLSL Translator + GL vertex shader/compute shader
PICA geometry shader -> ??? (fall back to CPU?)
PICA primitive assembly, quaternion fix, clipping -> GL geometry shader
* PICA Vertex loader -> GL vertex loader+vertex shader

tywald commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+It was a dark time where [Fire Emblem was mostly unplayable](https://github.com/citra-emu/citra/issues/2548) and [Pokémon didn't have attack animations](https://github.com/citra-emu/citra/issues/2629).
+Geometry shaders is a GPU stage where the game can transform a single primitive into zero or more primitives.
+This can be used to create particle effects from a few points or can be used to fill in gaps in a 3D model.
+Citra has supported Geometry Shaders since the summer of 2016 thanks to the heroic efforts of the many people that have worked on it in the past.
```
GS was merged in 2017.

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+Very fast. Test results across various computers show that it **averages out to be a 2x speed boost.**
+With the new update, Citra will use much more of your GPU, removing some of the dependence on a CPU with high single-core performance.
+(As always, the actual difference will vary by game and by your specific hardware configuration!)
```
why parenthesis?

BreadFish64 commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+No one had yet tested to see if it worked on AMD GPUs.
+We called for our good friend [JMC47](https://dolphin-emu.org/blog/authors/JMC47/) to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! Oh no!
+
+From JMC47's time in Dolphin, he's made a few friends here and there, and he found someone willing to investigate.
```
`From JMC47's time in Dolphin, he's made a few friends here and there`
JMC47 has made a few friends here and there from his time in Dolphin.
`and he found someone`
could probably remove he

Schplee replied:
I feel @jroweboy's original sentence is less awkward than "JMC47 has made a few friends here and there from his time in Dolphin."

The article utilizes a casual conversation tone, and that tone doesn't translate with your change.

BreadFish64 replied:
I see your point, but using a pronoun as a subject with the antecedent in the same sentence is kind of awkward. Another way he could make the sentence less clunky is by maybe removing `here and there` and changing `from` to 'during'

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+## Moving Forward with the Hardware Renderer: Cleaning up Texture Forwarding
+
+Despite the issues mentioned above, OpenGL has been a fair choice for a hardware renderer, and [phantom](https://github.com/phanto-m/) has been hard at work making improving the renderer.
```
`has been hard at work making improving the renderer.`

Should be `has been hard at work improving the renderer.`

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+The communication between CPU and GPU is optimized for large amounts of data transferred, but as a tradeoff, it's not very fast.
+This works great for PC games, where you know all the textures you want to upload ahead of time and can send them in one large batch, but ends up hurting performance for Citra since we can't know in advance when the emulated game will do something that requires a texture upload.
+
+The texture forwarding rewrite increases the speed of many games by adding in new checks to avoid this costly syncronization of textures between emulated 3DS memory and the host GPU memory.
```
`syncronization` should be `synchronization`

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+This works great for PC games, where you know all the textures you want to upload ahead of time and can send them in one large batch, but ends up hurting performance for Citra since we can't know in advance when the emulated game will do something that requires a texture upload.
+
+The texture forwarding rewrite increases the speed of many games by adding in new checks to avoid this costly syncronization of textures between emulated 3DS memory and the host GPU memory.
+Additionally, the new texture forwarding can avoid even more texture uploads by copying the data from an existing and compatible locations.
```
`copying the data from an existing and compatible locations.`

`an` should be removed

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+The texture forwarding rewrite increases the speed of many games by adding in new checks to avoid this costly syncronization of textures between emulated 3DS memory and the host GPU memory.
+Additionally, the new texture forwarding can avoid even more texture uploads by copying the data from an existing and compatible locations.
+As an extension of this feature, phantom went the extra mile and fixed Pokémon outlines as well!
+Pokémon games would draw the outline by reinterpreting the depth and stencil buffer as a RGBA texture, using the value for the red color to draw the outline.
```
`stencil buffer as a RGBA texture`

`a` should be `an`

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+### Multiplication Shouldn't Be This Slow
+
+When converting from PICA200 shaders into GLSL, there's a few PICA200 opcodes that should just match up without any issues.
```
`there's a few` should be `there are a few`

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+    "accurate_mul_on.png=On: I don't see any issue..." 
+    "accurate_mul_off.png=Off: Oh. Well maybe Eren looks better this way" >}}
+
+It turns out that the PICA200 multiplication opcode has a few edge cases that doesn't impact a large majority of games, and leads to some hilarious results in others.
```
`has a few edge cases that doesn't impact`

`doesn't` should be `don't`

Schplee commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+### Finding Bugs and Working Overtime
+
+We were very excited to launch this feature when phantom declared that it was ready; results from user testing was entirely positive, and the performance improvements were unbelievable, but one thing stood in the way.
```
`results from user testing was entirely`

`was` should be `were`

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+>}}
+
+As a general rule of thumb, CPUs are fast at computing general tasks, while GPUs are *blazing fast* at computing very specific tasks.
+Whenever the tasks the PICA200 performs matches well with tasks you can do on a GPU using OpenGL, everything is fast and everyone is happy!
```
`performs matches`? engrish

bunnei replied:
`performs matches`? engrish

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+As a general rule of thumb, CPUs are fast at computing general tasks, while GPUs are *blazing fast* at computing very specific tasks.
+Whenever the tasks the PICA200 performs matches well with tasks you can do on a GPU using OpenGL, everything is fast and everyone is happy!
+That said, we tend to run into edge cases that the PICA200 supports, but frankly, OpenGL is not well suited to support.
+This leads to cases where sometimes we just have to [live with minor inaccuracies as a tradeoff for speed](https://github.com/citra-emu/citra/pull/2697)
```
Missing punctuation

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+{{< figure src="/images/entry/improvements-to-hardware-renderer/outline_bug.png" title="The infamous Pokémon outlines bug was one such example of something that OpenGL just doesn't handle well" >}}
+
+OpenGL is also great for emulator developers because it's a cross-platform standard for graphics, with support for all major desktop platforms.
+But because it's a standard, this means performance and features can vary widely between operating systems, graphics driver, and the physical graphics card.
```
Because it's a standard is the reason performance and features vary? Can we give better reasons than that?

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+The texture forwarding rewrite increases the speed of many games by adding in new checks to avoid this costly syncronization of textures between emulated 3DS memory and the host GPU memory.
+Additionally, the new texture forwarding can avoid even more texture uploads by copying the data from an existing and compatible locations.
+As an extension of this feature, phantom went the extra mile and fixed Pokémon outlines as well!
```
nit: IMO, all username references should be linked. Makes the sentences read less weird without capitalization

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+But before we get into why this was so challenging, let's see some real performance numbers!
+
+Average performance increase
+graph: performance of dedicated cards vs integrated
```
Missing punctuation? e.g. a colon or something?

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+graph: performance of dedicated cards vs integrated
+
+According to telemetry data, Pokémon is the most played game on Citra
+graph: performance of sun/moon oras
```
I assume we are missing graphics here... will stop commenting :P

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+When trying to convert the PICA200 geometry shaders into GLSL, phantom ran into some very tough problems.
+Some games rely on the shader state carrying over from run to run, which isn't something that you can do in GLSL.
+Even more so, some games rely on (i know i'm forgetting something here??????).
```
remove

jroweboy replied:
I really wanted to add a segment on geometry shaders, but i honestly don't know enough about the challenges of supporting them in GLSL so I'll just remove the entire paragraph. We can re add it if someone who knows a thing or two about them chimes in before we go live.

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+Alas, it's a performance penalty to use a function everywhere instead of regular multiplication.
+On weaker GPUs, we noticed the penalty is so severe, we actually made this configurable.
+The whole point of a hardware renderer is to be fast, so eating a penalty when only a small handful of games need this level of accuracy would be regrettable.
+You can turn off this feature in the settings by deselecting "Accurate Shader" and get a noticeable performance boost, but be aware that a few games will break in strange ways!
```
Should we show a screenshot of the config UI with the option highlighted? So this can also self serve as documentation

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+We called for our good friend [JMC47](https://dolphin-emu.org/blog/authors/JMC47/) to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! Oh no!
+
+From JMC47's time in Dolphin, he's made a few friends here and there, and he found someone willing to investigate.
+After a few grueling weeks, this developer was able to narrow down what the problem is, and luckily it's not a bug in the AMD drivers.
```
Is it worth naming the dev who helped us out to thank them?

jroweboy replied:
i just barely learned who it was today ^^;

bunnei commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,200 @@
+
+It's a challenge to support all of the many GPUs out there, and we've put in so much work to ensure that this new feature will run on as many hardware configurations as possible.
+But it's very likely that there will be some GPUs that do not fully support the new hardware renderer, and so we added another option in the Configuration to allow users to turn this feature off entirely.
+Simply changing the "Hardware Renderer" from "GPU Shaders" to "CPU Shaders" will revert to using the same old CPU JIT shaders that Citra was using before.
```
Worth stating what the new default behavior will be, IMO
Also, worth adding a screenshot of the config as well with the setting highlighted, IMO

jroweboy commented:
Heres a preview of what the front page looks like with the banner: 
![firefox_2018-03-09_23-52-29](https://user-images.githubusercontent.com/952515/37239245-4e95b7c0-23f5-11e8-982d-eeaf3154a50b.png)

Note that both of the pokemon pictures are mp4 autoplay looping video (aka they look like high quality gifs)

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+
+## The Big News You've Been Waiting For
+
+A few paragrahs ago, we mentioned that Citra's hardware renderer did most of the emulation on the CPU, and only some of it on the GPU.
```
s/paragrahs/paragraphs/

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+{{< figure src="/images/entry/improvements-to-hardware-renderer/outline_bug.png" title="The infamous Pokémon outlines bug was one such example of something that OpenGL just doesn't handle well" >}}
+
+OpenGL is also great for emulator developers because it's a cross-platform standard for graphics, with support for all major desktop platforms.
+But because OpenGL is just a specification, every vendor is left up to their own to make their drivers support the specification for every individual platorm.
```
s/platorm/platform/

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+{{< figure type="youtube" id="sYukdJam6gk" title="It took a year from when citra started to get to this point" >}}
+
+This engineering feat was thanks to the hardwork of many contibutors in both the emulator scene and the 3ds hacking scene, who worked tirelessly to reverse engineer the 3DS GPU, a chip called the PICA200.
+But not even a few months later, Citra was able to play the game at fullspeed! 
```
s/fullspeed/full speed/

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+
+{{< figure type="youtube" id="sYukdJam6gk" title="It took a year from when citra started to get to this point" >}}
+
+This engineering feat was thanks to the hardwork of many contibutors in both the emulator scene and the 3ds hacking scene, who worked tirelessly to reverse engineer the 3DS GPU, a chip called the PICA200.
```
s/hardwork/hard work/
s/contibutors/contributors/

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+The PICA200 supports barebones `CALL`, `IF`, and `LOOP` operations, but also supports an arbitrary `JMP` that can go to any address.
+Translating PICA200 shaders into GLSL (OpenGL Shader Language) means that you'll have to be prepared to rewrite every arbitrary `JMP` without using a `goto` as GLSL doesn't support them.
+
+[phantom](https://github.com/phanto-m/) assumed the worst when they originally translated PICA200 shaders into GLSL and wrote a monsterous switch statement that would have a case for every jump target and act as a PICA200 shader interpreter.
```
s/monsterous/monstrous/

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+This worked, but proved to be slower than the software renderer!
+Now that [phantom](https://github.com/phanto-m/) knew it was possible, and they had some data about how the average PICA200 shader looked, they took to rewrite it with the goal to make it fast.
+While the shaders could theoretically be very unruly and hard to convert, almost all the shaders were well behaved, presumably because they are compiled from a higher level language.
+This time around, [phantom](https://github.com/phanto-m/) generated native GLSL functions wherever possible by analyizing the control flow of the instructions, and the results are much prettier and faster.
```
s/analyizing/analyzing/

MerryMage commented on site/content/entry/improvements-to-hardware-renderer.md:
```diff
@@ -0,0 +1,192 @@
+We called for our good friend [JMC47](https://dolphin-emu.org/blog/authors/JMC47/) to break out the AMD card he uses for testing Dolphin, and Citra crashed the driver! Oh no!
+
+From [JMC47](https://dolphin-emu.org/blog/authors/JMC47/)'s time in Dolphin, he's made a few friends here and there, and he found someone willing to investigate.
+After a few grueling weeks, [JonnyH](https://github.com/JonnyH) was able to narrow down what the problem is, and luckily it's not a bug in the AMD drivers.
```
s/grueling/gruelling/
